engines/sci/util.h

```diff
@@ -45,6 +50,715 @@ void WRITE_SCI11ENDIAN_UINT32(void *ptr, uint32 val);
+
+		void *buffer = malloc(readBytes);
+		assert(buffer != nullptr);
+		assert(stream.read(buffer, readBytes) == readBytes);
+
+		_span.data = (value_type *)buffer;
+		_span._size = readBytes / sizeof(value_type);
+		_span._sourceName = sourceName;
+		_span._sourceByteOffset = position;
+	}
+
+	inline void allocateFromFile(Common::File &file, size_type readBytes = kSpanMaxSize) {
+		allocateFromStream(file, readBytes, file.getName());
+	}
+
+	/**
+	 * Transfers ownership of the Span from the other owner to this owner.
+	 * If this owner already holds another Span, the old Span will be destroyed.
+	 */
+	// TODO: Maybe delete before assign should be explicit?
+	inline SpanOwner &operator=(SpanOwner &other) {
+		if (_span.data) {
+			free((void *)_span.data);
+		}
+		_span = other._span;
+		other.release();
+		return *this;
+	}
+
+	inline operator bool() const { return _span; }
+
+	inline const Span &operator*() const { return _span; }
+	inline Span &operator*() { return _span; }
+
+	inline const Span *operator->() const { return &_span; }
+	inline Span *operator->() { return &_span; }
+
+	inline const value_type &operator[](const index_type index) const { return _span[index]; }
+	inline value_type &operator[](const index_type index) { return _span[index]; }
+
+	/**
+	 * Releases the pointer owned by this owner to the caller.
+	 */
+	inline value_type *release() {
+		value_type *data = _span.data;
+		_span.clear();
+		return data;
+	}
+
+	/**
+	 * Destroys the memory owned by this owner.
+	 */
+	inline void clear() {
+		free((void *)_span.data);
+		_span.clear();
+	}
+
+private:
+	// TODO: Would it be better to store the memory address here? Then spans
+	// would be free to implement arithmetic operations that get them to act
+	// more like pointers.
+	Span _span;
+};
+
+#pragma mark -
+#pragma mark Span
+
+/**
+ * Bounds-checked random access into a contiguous block of memory.
+ */
+template <typename ValueType>
+class Span {
+	friend class SpanInternal::SpanIterator<Span, true>;
+	friend class SpanInternal::SpanIterator<Span, false>;
+#if !defined(__GNUC__) || GCC_ATLEAST(3, 0)
+	template <typename T> friend class SpanOwner;
+	template <typename T> friend class Span;
+#endif
+
+public:
+	typedef ValueType value_type;
+	typedef uint32 index_type;
+	typedef uint32 size_type;
+	typedef SpanInternal::SpanIterator<Span, true> const_iterator;
+	typedef SpanInternal::SpanIterator<Span, false> iterator;
+
+	inline Span() :
+		data(nullptr),
+		_size(0),
+		_sourceName(),
+		_sourceByteOffset(0) {}
+
+	inline Span(value_type *data_,
+				const size_type dataSize,
+				const Common::String &sourceName = Common::String(),
+				const index_type sourceByteOffset = 0) :
+		data(data_),
+		_size(dataSize),
+		_sourceName(sourceName),
+		_sourceByteOffset(sourceByteOffset) {}
+
+	template <typename OtherValueType>
+	inline Span(const Span<OtherValueType> &other) {
+		data = (typename RemoveConst<value_type>::type *)other.data;
```

Comment:
Unless I misunderstand, this RemoveConst is a bit problematic since it allows implicit casts from `Span<const T>` to `Span<T>`, which seems dangerous. This should probably just be `data = other.data;`

The one essential place where this is used seems to be in `ScriptPatcher::applyPatch`. (And one less essential place in `GfxView::getBitmap`)